crates/core/src/checkpoint.rs

```diff
@@ -164,6 +245,40 @@ impl CheckPoint {
             })
     }
 
+    /// This method tests for `self` and `other` to have equal internal pointers.
+    pub fn eq_ptr(&self, other: &Self) -> bool {
+        Arc::as_ptr(&self.0) == Arc::as_ptr(&other.0)
+    }
+}
+
+impl<B> CheckPoint<B>
+where
+    B: Copy + core::fmt::Debug + ToBlockHash + From<BlockHash>,
```

Comment:
```suggestion
    B: Copy + core::fmt::Debug + ToBlockHash,
```

Why do we need `From<BlockHash>`?